client/batch.go

```diff
@@ -0,0 +1,384 @@
+	for _, r := range b.Results {
+		if err := r.Err; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *Batch) initResult(calls, numRows int, err error) {
+	r := Result{calls: calls, Err: err}
+	if numRows > 0 {
+		if b.rowsIdx+numRows <= len(b.rowsBuf) {
+			r.Rows = b.rowsBuf[b.rowsIdx : b.rowsIdx+numRows]
+			b.rowsIdx += numRows
+		} else {
+			r.Rows = make([]KeyValue, numRows)
+		}
+	}
+	if b.Results == nil {
+		b.Results = b.resultsBuf[0:0]
+	}
+	b.Results = append(b.Results, r)
+}
+
+func (b *Batch) fillResults() error {
+	offset := 0
+	for i := range b.Results {
+		result := &b.Results[i]
+
+		for k := 0; k < result.calls; k++ {
+			call := b.calls[offset+k]
+
+			if result.Err == nil {
+				result.Err = call.Reply.Header().GoError()
+			}
+
+			switch t := call.Reply.(type) {
+			case *proto.GetResponse:
+				row := &result.Rows[k]
+				row.Key = []byte(call.Args.(*proto.GetRequest).Key)
+				if result.Err == nil {
+					row.setValue(t.Value)
+				}
+			case *proto.PutResponse:
+				req := call.Args.(*proto.PutRequest)
+				row := &result.Rows[k]
+				row.Key = []byte(req.Key)
+				if result.Err == nil {
+					row.setValue(&req.Value)
+					row.setTimestamp(t.Timestamp)
+				}
+			case *proto.ConditionalPutResponse:
+				req := call.Args.(*proto.ConditionalPutRequest)
+				row := &result.Rows[k]
+				row.Key = []byte(req.Key)
+				if result.Err == nil {
+					row.setValue(&req.Value)
+					row.setTimestamp(t.Timestamp)
+				}
+			case *proto.IncrementResponse:
+				row := &result.Rows[k]
+				row.Key = []byte(call.Args.(*proto.IncrementRequest).Key)
+				if result.Err == nil {
+					// TODO(pmattis): Should IncrementResponse contain a
+					// proto.Value so that the timestamp can be returned?
+					row.Value = &t.NewValue
+				}
+			case *proto.ScanResponse:
+				result.Rows = make([]KeyValue, len(t.Rows))
+				for j, kv := range t.Rows {
+					row := &result.Rows[j]
+					row.Key = kv.Key
+					row.setValue(&kv.Value)
+				}
+			case *proto.DeleteResponse:
+				row := &result.Rows[k]
+				row.Key = []byte(call.Args.(*proto.DeleteRequest).Key)
+
+			case *proto.AdminMergeResponse:
+			case *proto.AdminSplitResponse:
+			case *proto.DeleteRangeResponse:
+			case *proto.EndTransactionResponse:
+			case *proto.InternalBatchResponse:
+			case *proto.InternalGCResponse:
+			case *proto.InternalMergeResponse:
+			case *proto.InternalPushTxnResponse:
+			case *proto.InternalRangeLookupResponse:
+			case *proto.InternalResolveIntentResponse:
+			case *proto.InternalResolveIntentRangeResponse:
+				// Nothing to do for these methods as they do not generate any
+				// rows. For the proto.Internal* responses the caller will have hold of
+				// the response object itself.
+
+			default:
+				if result.Err == nil {
+					result.Err = fmt.Errorf("unsupported reply: %T", call.Reply)
+				}
+			}
+		}
+		offset += result.calls
+	}
+
+	for i := range b.Results {
+		result := &b.Results[i]
+		if result.Err != nil {
```

Comment:
Yes, but the savings is minor and this code was only moved `db.go`. I'm going to leave as-is for now.